### PR TITLE
General update

### PR DESCRIPTION
**Pull request by Sparkycz**
Hey man,

I had implemented a very similar async cache using Redis as storage in a project. 
I applied some changes to merge my implementation with your one. Check it out. Thanks.

- Updated backends - Serialization/Deserialization of values moved to one point
- Updated backends - Implementation of PEP 3119
- Updated backends - Fix cache.get_or_set called with coroutine
- Updated core - Add caching coroutine easily by `__call__()`

### CODE REVIEW

**Comment by Sparkycz** [comment]
@rafalp please, check it out

**Comment by rafalp** [comment]
Looks good. I'm only wondering if type shouldn't be `Union[Awaitable[Serializable], Serializable]`?

**Comment by Sparkycz** [comment]
Is `typing` able to check it? I don't know How exact should it be.. However, I think `Awaitable` is enough..

**Comment by rafalp** [comment]
@Sparkycz mypy checks this. `Awaitable` is a generic type just like `List` or `Dict` are.

**Comment by Sparkycz** [comment]
OK, updated.

File: caches/backends/base.py [hunk 1]
```diff
@@ -1,56 +1,97 @@
-from typing import Any, Dict, Iterable, Mapping, Optional, Union
+from abc import ABCMeta, abstractmethod
+from typing import Any, Awaitable, Dict, Iterable, Mapping, Optional, Union
+
+import json
```
Comment (rafalp) [inline caches/backends/base.py]: This line shoud be placed before `from abc`.

File: caches/core.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 from types import TracebackType
-from typing import Any, Dict, Iterable, Mapping, Optional, Type, Union
+from typing import Any, Awaitable, Coroutine, Dict, Iterable, Mapping, Optional, Type, Union
 from urllib.parse import SplitResult, parse_qsl, urlsplit
 
+import json
```
Comment (rafalp) [inline caches/core.py]: This line should be before `from types`

**Review by rafalp — CHANGES_REQUESTED** [review]

**Comment by rafalp** [comment]
Found two more things, but those are small enough. Rest is looking good to me. Thanks!

**Review by rafalp — APPROVED** [review]

**Pull request merged**